samtranslator/model/apigateway.py

```diff
@@ -85,3 +86,142 @@ def make_auto_deployable(self, stage, swagger=None):
         hash = generator.get_hash(length=40)  # Get the full hash
         self.Description = "RestApi deployment id: {}".format(hash)
         stage.update_deployment_ref(self.logical_id)
+
+
+class ApiGatewayAuthorizer(object):
+    def __init__(self, api_logical_id=None, name=None, user_pool_arn=None, function_arn=None, identity=None,
+                 function_payload_type=None, function_invoke_role=None):
+        self.name = name
+        self.user_pool_arn = user_pool_arn
+        self.function_arn = function_arn
+        self.identity = identity
+        self.function_payload_type = function_payload_type
+        self.function_invoke_role = self._construct_role(function_invoke_role)
+
+    def generate_swagger(self):
+        authorizer_type = self._get_type()
+        APIGATEWAY_AUTHORIZER_KEY = 'x-amazon-apigateway-authorizer'
+        swagger = {
+            "type": "apiKey",
+            "name": self._get_swagger_header_name(),
+            "in": "header",
+            "x-amazon-apigateway-authtype": self._get_swagger_authtype(),
+            "x-amazon-apigateway-authorizer": {
+                "type": self._get_swagger_authorizer_type()
+            }
+        }
+
+        if authorizer_type == 'COGNITO_USER_POOLS':
+            swagger[APIGATEWAY_AUTHORIZER_KEY]['providerARNs'] = self._get_user_pool_arn_array()
+
+        elif authorizer_type == 'LAMBDA':
+            partition = ArnGenerator.get_partition_name()
+            authorizer_uri = fnSub('arn:{0}:apigateway:${{AWS::Region}}:lambda:path/2015-03-31/functions/'
+                                   '${{__FunctionArn__}}/invocations'.format(partition),
+                                   {'__FunctionArn__': self.function_arn})
+
+            swagger[APIGATEWAY_AUTHORIZER_KEY]['authorizerUri'] = authorizer_uri
+            reauthorize_every = self._get_reauthorize_every()
+            function_invoke_role = self._get_function_invoke_role()
+
+            if reauthorize_every:
+                swagger[APIGATEWAY_AUTHORIZER_KEY]['authorizerResultTtlInSeconds'] = reauthorize_every
+
+            if function_invoke_role:
+                swagger[APIGATEWAY_AUTHORIZER_KEY]['authorizerCredentials'] = function_invoke_role
+
+            if self._get_function_payload_type() == 'REQUEST':
+                swagger[APIGATEWAY_AUTHORIZER_KEY]['identitySource'] = self._get_identity_source()
+
+        if authorizer_type == 'COGNITO_USER_POOLS' or (authorizer_type == 'LAMBDA' and self._get_function_payload_type() == 'TOKEN'):
+            identity_validation_expression = self._get_identity_validation_expression()
+
+            if identity_validation_expression:
+                swagger[APIGATEWAY_AUTHORIZER_KEY]['identityValidationExpression'] = identity_validation_expression
+
+        return swagger
+
+    def _construct_role(self, function_invoke_role):
```

Comment:
Rename to `_get_function_invoke_role`. Initially planned on allowing users to specify just a role name and we would then construct a full ARN if one wasn't provided. I don't plan on doing this for initial release; best practice is to define a role in the template (or send in via Parameters, etc.)

Reply:
Removed. `_get_function_invoke_role` already exists.